Sources/SwiftFusion/Inference/BearingRangeFactor.swift

```diff
@@ -0,0 +1,193 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+//
+///// Protocol for indicating a struct is range-capable
+//public protocol HasRange: Differentiable {
+//  associatedtype A: Differentiable
+//  associatedtype B: Differentiable where B.TangentVector: EuclideanVectorSpace & VectorConvertible & AdditiveArithmetic
+//  @differentiable
+//  static func range(_ from: A, _ to: B) -> Double
+//}
+//
+//public struct PoseToPointRange2D: HasRange {
+//  @differentiable
+//  public static func range(_ from: Pose2, _ to: Vector2) -> Double {
+//    (to - from.t).norm
+//  }
+//}
+//
+//public struct PointToPointRange2D: HasRange {
+//  @differentiable
+//  public static func range(_ from: Vector2, _ to: Vector2) -> Double {
+//    (to - from).norm
+//  }
+//}
+//
+///// Protocol for indicating a struct is bearing-capable
+//public protocol HasBearing: Differentiable {
+//  associatedtype A: LieGroup where A.TangentVector: VectorConvertible
+//  associatedtype B: Differentiable where B.TangentVector: EuclideanVectorSpace & VectorConvertible & AdditiveArithmetic
+//  associatedtype C: EuclideanVectorSpace & Differentiable & AdditiveArithmetic & VectorProtocol & TangentStandardBasis
+//  @differentiable
+//  static func bearing(_ from: A, _ to: B) -> C
+//}
+//
+//public struct PoseToPointBearing2D: HasBearing {
+//  @differentiable
+//  public static func bearing(_ from: Pose2, _ to: Vector2) -> Vector1 {
+//    let dx = (to - from.t)
+//    return Vector1(between(Rot2(c: dx.x / dx.norm, s: dx.y / dx.norm), from.rot).theta)
+//  }
+//}
+//
+//public struct BearingRangeError<BE: TangentStandardBasis, RE: TangentStandardBasis>:
+//  Differentiable & KeyPathIterable & TangentStandardBasis & VectorConvertible {
+//  @differentiable
+//  public init(_ vector: Vector) {
+//
+//  }
+//
+//  public var vector: Vector
+//
+//  public var bearing: BE
+//  public var range: RE
+//}
+//
+///// A `NonlinearFactor` that calculates the bearing and range error of one pose and one landmark
+/////
+///// Input is a dictionary of `Key` to `Value` pairs, and the output is the scalar
+///// error value
+/////
+///// Interpretation
+///// ================
+///// `Input`: the input values as key-value pairs
+/////
+//public struct BearingRangeFactor<B: HasBearing, R: HasRange>: NonlinearFactor
```

Comment:
ideally this should be `BearingRangeFactor<Base, Target>` but I have no idea how to implement this...

Reply:
With the simplification I mentioned above, you can get pretty close:

```swift
public struct BearingRangeFactor<BearingRangeFunction: BearingFunction>: NonlinearFactor
  /* where ... */
{
  public typealias Base = BearingRangeFunction.Base
  public typealias Target = BearingRangeFunction.Target
}
```

With this, the implementation of the `BearingRangeFactor` can use `Base` and `Target` types everywhere.

Clients who instantiate `BeringRangeFactor`s will need to find the right type to use, e.g. `BearingRangeFactor<PoseToPoint2D>(...)`. This isn't quite as good as `BearingRangeFactor<Pose2, Point2>(...)`, but I think it's pretty close.

I don't think there's a way in Swift to automatically find the right "trait" based on multiple types.